File: backend/mcapi/user/notes.py
from ..mcapp import app
from ..decorators import jsonp, apikey, eventlog
import rethinkdb as r
from flask import g, request
from .. import dmutil, error, resp
from loader.model import note


@app.route('/notes/<project_id>', methods=['GET'])
@jsonp
def get_notes(project_id):
    notes = list(r.table('notes')
                 .get_all(project_id, index='project_id')
                 .run(g.conn))
    return resp.to_json(notes)


@app.route('/notes', methods=['POST'])
@apikey(shared=True)
@eventlog
def add_note():
    j = request.get_json()
    item_id = dmutil.get_required('item_id', j)
    item_type = dmutil.get_required('item_type', j)
    creator = dmutil.get_required('creator', j)
    title = dmutil.get_required('title', j)
    message = dmutil.get_required('note', j)
    project_id = dmutil.get_required('project_id', j)
    n = note.Note(creator, message, title, item_id, item_type, project_id)
    created_note = dmutil.insert_entry('notes', n.__dict__,
                                       return_created=True)
    return resp.to_json(created_note)


@app.route('/notes', methods=['PUT'])
@apikey(shared=True)
@jsonp
@eventlog
def update_note():
    j = request.get_json()
    title = dmutil.get_optional('title', j)
    message = dmutil.get_optional('note', j)
    note_id = dmutil.get_required('id', j)
    if message or title:
        result = r.table('notes').get(note_id).update(
            {'title': title, 'note': message,'mtime': r.now()},
            return_changes=True).run(g.conn, time_format='raw')
        updated_note = result['changes'][0]['new_val']
        return resp.to_json(updated_note)
    else:
        return error.not_acceptable("Unable to update note: " + note_id)

File: backend/servers/mcapi/db/model/notes.js
const r = require('../r');
const model = require('./model');
const db = require('./db');

function* createNote(note, itemType, itemId) {
    let note = yield db.insert('notes', note);
    let n2i = new model.Note2Item(itemId, itemId, note.id);
    yield db.insert('note2item', n2i);
    return note;
}

function* updateNote(note, noteId) {
    note.mtime = r.now();
    yield r.table('notes').get(noteId).update(note);
}

function* updateNotes(notes) {
    for (let note of notes) {
        note.mtime = r.now();
    }

    yield r.table('notes').insert(notes, {conflict: 'update'});
}

function* deleteNote(noteId) {
    yield r.table('notes').get(noteId).delete();
    yield r.table('note2item').getAll(noteId, {index: 'note_id'}).delete();
}

function* getAllNotesForItem(itemId) {
    return yield r.table('note2item')
        .getAll(itemId, {index: 'item_id'}).without('item_id', 'item_type')
        .eqJoin('note_id', r.table('notes')).without({left: {note_id: true}}).zip();
}

function getAllNotesForItemRQL(itemId) {
    return r.table('note2item')
        .getAll(itemId, {index: 'item_id'}).without('item_id', 'item_type')
        .eqJoin('note_id', r.table('notes')).without({left: {note_id: true}}).zip();
}

function* getNote(noteId) {
    return yield r.table('notes').get(noteId);
}

module.exports = {
    createNote,
    updateNote,
    updateNotes,
    deleteNote,
    getAllNotesForItem,
    getAllNotesForItemRQL,
    getNote,
};